mono/metadata/object.c

```diff
@@ -7580,88 +7580,24 @@ mono_message_init (MonoDomain *domain,
 {
 	MONO_REQ_GC_UNSAFE_MODE;
 
-	static MonoClass *object_array_klass;
-	static MonoClass *byte_array_klass;
-	static MonoClass *string_array_klass;
-	mono_error_init (error);
-	MonoMethodSignature *sig = mono_method_signature (method->method);
-	MonoString *name;
-	MonoArray *arr;
-	int i, j;
-	char **names;
-	guint8 arg_type;
-
-	if (!object_array_klass) {
-		MonoClass *klass;
+	static MonoMethod *init_message_method = NULL;
 
-		klass = mono_array_class_get (mono_defaults.byte_class, 1);
-		g_assert (klass);
-		byte_array_klass = klass;
-
-		klass = mono_array_class_get (mono_defaults.string_class, 1);
-		g_assert (klass);
-		string_array_klass = klass;
-
-		klass = mono_array_class_get (mono_defaults.object_class, 1);
-		g_assert (klass);
-
-		mono_atomic_store_release (&object_array_klass, klass);
+	if (!init_message_method) {
+		init_message_method = mono_class_get_method_from_name (mono_defaults.mono_method_message_class, "InitMessage", 2);
```

Comment:
@akoeplinger two reasons:

* until recently they were not open source;
* each runtime has some special cases, e.g. SRE is not needed for XI...

Both are solvable - matter of investing the time to merge it (without breaking any SDK).

As for XML mscorlib is an exception (for XI/XM) and all other similar requirements are done by code for other assemblies. Again two reasons:

* mscorlib is the biggest one;
* mscorlib is more _basic_ (most of them are unconditionally needed).

One of the main reason for doing it by code(for other assemblies) was to add a layer of custom checks to see if the types/methods were really needed. E.g. you don't want to preserve internal Socket helpers if your code does not use them (the XML syntax does not allow that level of control).